crates/astria-sequencer/src/transaction/mod.rs

```diff
@@ -1,9 +1,131 @@
+    SignedTransaction,
+    UnsignedTransaction,
+};
+use tracing::instrument;
+
+use crate::accounts::state_ext::{
+    StateReadExt,
+    StateWriteExt,
+};
+
+pub(crate) fn check_stateless(tx: &SignedTransaction) -> anyhow::Result<()> {
+    tx.unsigned_transaction().check_stateless()
+}
+
+pub(crate) async fn check_stateful<S: StateReadExt + 'static>(
+    tx: &SignedTransaction,
+    state: &S,
+) -> anyhow::Result<()> {
+    let signer_address = Address::from_verification_key(tx.verification_key());
+    tx.unsigned_transaction()
+        .check_stateful(state, signer_address)
+        .await
+}
+
+pub(crate) async fn execute<S: StateWriteExt>(
+    tx: &SignedTransaction,
+    state: &mut S,
+) -> anyhow::Result<()> {
+    let signer_address = Address::from_verification_key(tx.verification_key());
+    tx.unsigned_transaction()
+        .execute(state, signer_address)
+        .await
+}
+
+#[async_trait::async_trait]
+impl ActionHandler for UnsignedTransaction {
+    fn check_stateless(&self) -> anyhow::Result<()> {
+        ensure!(!self.actions.is_empty(), "must have at least one action");
+
+        for action in &self.actions {
+            match action {
+                Action::Transfer(act) => act
+                    .check_stateless()
+                    .context("stateless check failed for TransferAction")?,
+                Action::Sequence(act) => act
+                    .check_stateless()
+                    .context("stateless check failed for SequenceAction")?,
+            }
+        }
+        Ok(())
+    }
+
+    async fn check_stateful<S: StateReadExt + 'static>(
+        &self,
+        state: &S,
+        from: Address,
+    ) -> anyhow::Result<()> {
+        // Nonce should be equal to the number of executed transactions before this tx.
+        // First tx has nonce 0.
+        let curr_nonce = state.get_account_nonce(from).await?;
+        ensure!(
+            curr_nonce == self.nonce,
+            "invalid nonce, tx nonce must match account nonce"
+        );
+
+        // do we need to make a StateDelta here so we can check the actions on the successive state?
+        for action in &self.actions {
+            match action {
+                Action::Transfer(act) => act
+                    .check_stateful(state, from)
+                    .await
+                    .context("stateful check failed for TransferAction")?,
+                Action::Sequence(act) => act
+                    .check_stateful(state, from)
+                    .await
+                    .context("stateful check failed for SequenceAction")?,
+            }
+        }
+
+        Ok(())
+    }
+
+    #[instrument(
+        skip_all,
+        fields(
+            nonce = self.nonce,
+            from = from.to_string(),
+        )
+    )]
+    async fn execute<S: StateWriteExt>(&self, state: &mut S, from: Address) -> anyhow::Result<()> {
+        // TODO: make a new StateDelta so this is atomic / can be rolled back in case of error
```

Comment:
Is there an issue for this?

Reply:
This came over when I moved the code around. @noot: is this still up to date?

Maybe we can have a small PR that adds links to issues (not this PR though)

Reply:
https://github.com/astriaorg/astria/issues/317

Reply:
Added this to the PR